File: project_files/MSGrid.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace MineSnake
{
    public class MSGrid
    {
        private int height;
        private int width;
        private List<List<MSTile>> grid = new List<List<MSTile>>();

        public MSGrid()
        {
        }

        public MSGrid(int x, int y)
        {
            if (x <= 0 || y <= 0)
            {
                throw new MSException("SIZE_INSUFFICIENT");
            }

            height = x;
            width = y;

            for (int i = 0; i < x; i++)
            {
                List<MSTile> t = new List<MSTile>();

                for (int j = 0; j < y; j++)
                {
                    t.Add(new MSTile());
                }

                grid.Add(t);
            }
        }

        public int getHeight()
        {
            return height;
        }

        public int getWidth()
        {
            return width;
        }

        public List<List<MSTile>> getGrid()
        {
            return grid;
        }

        public void setTile(int x, int y, MSTile t)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            grid[x][y] = t;
        }

        public void setTile(int x, int y, MSTileProperty t)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            grid[x][y] = new MSTile(t);
        }

        public void setTile(int x, int y, List<MSTileProperty> ps)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            MSTile t = new MSTile();
            t.addProperties(ps);

            grid[x][y] = t;
        }

        public void setTiles(List<int> xs, List<int> ys, MSTileProperty p)
        {
            int xsize = xs.Count();
            int ysize = ys.Count();
            int l = xsize < ysize ? xsize : ysize;

            for (int i = 0; i < l; i++)
            {
                try
                {
                    setTile(xs[i], ys[i], p);
                }
                catch (MSException)
                {
                    throw new MSException("INDEX_OUT_OF_BOUNDS");
                }
            }
        }

        public void setTiles(List<int> xs, List<int> ys, List<MSTileProperty> ps)
        {
            int xsize = xs.Count();
            int ysize = ys.Count();
            int l = xsize < ysize ? xsize : ysize;

            for (int i = 0; i < l; i++)
            {
                setTile(xs[i], ys[i], ps);
            }
        }

        public MSTile getTile(int x, int y)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            return grid[x][y];
        }

        public void addProperty(int x, int y, MSTileProperty p)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            grid[x][y].addProperty(p);
        }

        public void addProperties(int x, int y, List<MSTileProperty> ps)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            grid[x][y].addProperties(ps);
        }

        public void addProperties(List<int> xs, List<int> ys, List<MSTileProperty> ps)
        {
            int xsize = xs.Count();
            int ysize = ys.Count();
            int l = xsize < ysize ? xsize : ysize;

            for (int i = 0; i < l; i++)
            {
                setTile(xs[i], ys[i], ps);
            }
        }

        public List<MSTileProperty> getProperties(int x, int y)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            return grid[x][y].getProperties();
        }

        public void removeProperty(int x, int y, MSTileProperty p)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            grid[x][y].removeProperty(p);
        }

        public bool isOf(int x, int y, MSTileProperty p)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            return grid[x][y].isOf(p);
        }

        public void decreaseMineCount(int x, int y)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            grid[x][y].decreaseMineCount();
        }

        public List<MSTile> getNeighbourhood(int x, int y)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            List<MSTile> t = new List<MSTile>();

            if (isIncluded(x - 1, y - 1))
            {
                t.Add(getTile(x - 1, y - 1));
            }
            if (isIncluded(x - 1, y))
            {
                t.Add(getTile(x - 1, y));
            }
            if (isIncluded(x - 1, y + 1))
            {
                t.Add(getTile(x - 1, y + 1));
            }
            if (isIncluded(x, y - 1))
            {
                t.Add(getTile(x, y - 1));
            }
            if (isIncluded(x, y + 1))
            {
                t.Add(getTile(x, y + 1));
            }
            if (isIncluded(x + 1, y - 1))
            {
                t.Add(getTile(x + 1, y - 1));
            }
            if (isIncluded(x + 1, y))
            {
                t.Add(getTile(x + 1, y));
            }
            if (isIncluded(x + 1, y + 1))
            {
                t.Add(getTile(x + 1, y + 1));
            }

            return t;
        }

        public List<Tuple<int, int>> getPassableNeighbourhood(int x, int y)
        {
            if (!isIncluded(x, y))
            {
                throw new MSException("INDEX_OUT_OF_BOUNDS");
            }

            List<Tuple<int, int>> t = new List<Tuple<int, int>>();

            if (isIncluded(x, y - 1) && !isOf(x, y - 1, MSTileProperty.OBSTACLE))
            {
                t.Add(new Tuple<int, int>(x + 0, y - 1));
            }
            if (isIncluded(x, y + 1) && !isOf(x, y + 1, MSTileProperty.OBSTACLE))
            {
                t.Add(new Tuple<int, int>(x + 0, y + 1));
            }
            if (isIncluded(x - 1, y) && !isOf(x - 1, y, MSTileProperty.OBSTACLE))
            {
                t.Add(new Tuple<int, int>(x - 1, y + 0));
            }
            if (isIncluded(x + 1, y) && !isOf(x + 1, y, MSTileProperty.OBSTACLE))
            {
                t.Add(new Tuple<int, int>(x + 1, y + 0));
            }

            return t;
        }

        public bool isIncluded(int x, int y)
        {
            return x >= 0 && y >= 0 && x < height && y < width;
        }

        public Tuple<int, int> getStart()
        {
            for (int i = 0; i < height; i++)
            {
                for (int j = 0; j < width; j++)
                {
                    if (isOf(i, j, MSTileProperty.START))
                    {
                        return new Tuple<int, int>(i, j);
                    }
                }
            }

            throw new MSException("NO_STARTING_TILE");
        }

        public void spawnFood()
        {
            Random r = new Random();

            while (true)
            {
                int xc = r.Next() % height;
                int yc = r.Next() % width;

                if (!isOf(xc, yc, MSTileProperty.OBSTACLE) && !isOf(xc, yc, MSTileProperty.START) && !isOf(xc, yc, MSTileProperty.MINE) && !isOf(xc, yc, MSTileProperty.SNAKE))
                {
                    addProperty(xc, yc, MSTileProperty.FOOD);
                    break;
                }
            }
        }
    }

}

File: project_files/cpp_version/MSGrid.cpp
#ifndef __GRID__
#define __GRID__

#include <vector>
#include <sstream>
#include <string>
#include <utility>
#include <cstdlib>
#include <ctime>
#include "MSTile.cpp"
#include "MSTileProperty.cpp"
#include "MSException.cpp"

// Represents a tile grid with methods of managing it.
class MSGrid {
private:
	int height;
	int width;
	std::vector<std::vector<MSTile>> grid;

public:
	// Creates an empty grid.
	MSGrid() {
	}

	// Creates a grid with the given size and fills it with empty tiles if it is accurately given; otherwise throws SIZE_INSUFFICIENT exception.
	MSGrid(int x, int y) {
		if (x <= 0 || y <= 0) {
			throw MSException::SIZE_INSUFFICIENT;
		}

		height = x;
		width = y;
		
		for (int i = 0; i < x; i ++) {
			std::vector<MSTile> t;

			for (int j = 0; j < y; j ++) {
				t.push_back(MSTile());
			}

			grid.push_back(t);
		}
	}

	// Returns the height of the grid.
	int getHeight() {
		return height;
	}

	// Returns the width of the grid.
	int getWidth() {
		return width;
	}

	// Sets a tile on a given coordinates to given tile if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	void setTile(int x, int y, MSTile t) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		grid[x][y] = t;
	}

	// Sets a tile on a given coordinates given property if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	void setTile(int x, int y, MSTileProperty t) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		grid[x][y] = t;
	}

	// Sets a tile on a given coordinates given properties if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	void setTile(int x, int y, const std::vector<MSTileProperty>& ps) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		grid[x][y] = MSTile(ps);
	}

	// Sets all the tiles on a given coordinates (pairing one-by-one with respect to the shorter vector) given property if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	void setTiles(const std::vector<int>& xs, const std::vector<int>& ys, MSTileProperty p) {
		int xsize = xs.size();
		int ysize = ys.size();
		int l = xsize < ysize ? xsize : ysize;

		for (int i = 0; i < l; i ++) {
			try {
				setTile(xs[i], ys[i], p);
			} catch (MSException) {
				throw MSException::INDEX_OUT_OF_BOUNDS;
			}
		}
	}

	// Sets all the tiles on a given coordinates (pairing one-by-one with respect to the shorter vector) given properties if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	void setTiles(const std::vector<int>& xs, const std::vector<int>& ys, const std::vector<MSTileProperty>& ps) {
		int xsize = xs.size();
		int ysize = ys.size();
		int l = xsize < ysize ? xsize : ysize;

		for (int i = 0; i < l; i ++) {
			setTile(xs[i], ys[i], ps);
		}
	}

	// Returns tile found on the given coordinates if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	MSTile getTile(int x, int y) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		return grid[x][y];
	}

	// Adds a property to a tile on given coordinates if it does not already contain one of the same kind; otherwise does nothing.
	// If coordinates are invalid, throws INDEX_OUT_OF_BOUNDS exception.
	void addProperty(int x, int y, MSTileProperty p) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		grid[x][y].addProperty(p);
	}

	// Adds all properties from the given vector to the tile on given coordinates ignoring ones that are already contained.
	// If coordinates are invalid, throws INDEX_OUT_OF_BOUNDS exception.
	void addProperties(int x, int y, const std::vector<MSTileProperty>& ps) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		grid[x][y].addProperties(ps);
	}

	// Adds all properties from the given vector to the tiles on given coordinates on given coordinates (pairing one-by-one with respect to the shorter vector) ignoring ones that are already contained.
	// If coordinates are invalid, throws INDEX_OUT_OF_BOUNDS exception.
	void addProperties(const std::vector<int>& xs, const std::vector<int>& ys, const std::vector<MSTileProperty>& ps) {
		int xsize = xs.size();
		int ysize = ys.size();
		int l = xsize < ysize ? xsize : ysize;

		for (int i = 0; i < l; i ++) {
			setTile(xs[i], ys[i], ps);
		}
	}

	// Returns vector containing all the properties of the tile on given coordinates if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	auto getProperties(int x, int y) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		return grid[x][y].getProperties();
	}

	// Removes given property from the tile on given coordinates properties if it exists; otherwise does nothing.
	// If coordinates are invalid, throws INDEX_OUT_OF_BOUNDS exception.
	void removeProperty(int x, int y, MSTileProperty p) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		grid[x][y].removeProperty(p);
	}

	// Checks whether the tile on given coordinates contains given property if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	bool isOf(int x, int y, MSTileProperty p) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		return grid[x][y].isOf(p);
	}

	// Decreases mine count of tile on give coordinates if those are correct; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	void decreaseMineCount(int x, int y) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		grid[x][y].decreaseMineCount();
	}

	// Returns vector consisting of tiles surrounding tile on a given coordinates if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	auto getNeighbourhood(int x, int y) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		std::vector<MSTile> t;

		if (isIncluded(x - 1, y - 1)) {
			t.push_back(getTile(x - 1, y - 1));
		}
		if (isIncluded(x - 1, y)) {
			t.push_back(getTile(x - 1, y));
		}
		if (isIncluded(x - 1, y + 1)) {
			t.push_back(getTile(x - 1, y + 1));
		}
		if (isIncluded(x, y - 1)) {
			t.push_back(getTile(x, y - 1));
		}
		if (isIncluded(x, y + 1)) {
			t.push_back(getTile(x, y + 1));
		}
		if (isIncluded(x + 1, y - 1)) {
			t.push_back(getTile(x + 1, y - 1));
		}
		if (isIncluded(x + 1, y)) {
			t.push_back(getTile(x + 1, y));
		}
		if (isIncluded(x + 1, y + 1)) {
			t.push_back(getTile(x + 1, y + 1));
		}

		return t;
	}

	// Returns vector consisting of pairs of coordinates surrounding tile on a given coordinates that can be accessed (ignoring placement of snake) if those are valid; otherwise throws INDEX_OUT_OF_BOUNDS exception.
	auto getPassableNeighbourhood(int x, int y) {
		if (!isIncluded(x, y)) {
			throw MSException::INDEX_OUT_OF_BOUNDS;
		}

		std::vector<std::pair<int, int>> t;

		if (isIncluded(x, y - 1) && !isOf(x, y - 1, MSTileProperty::OBSTACLE)) {
			t.push_back(std::make_pair<int, int>(x + 0, y - 1));
		}
		if (isIncluded(x, y + 1) && !isOf(x, y + 1, MSTileProperty::OBSTACLE)) {
			t.push_back(std::make_pair<int, int>(x + 0, y + 1));
		}
		if (isIncluded(x - 1, y) && !isOf(x - 1, y, MSTileProperty::OBSTACLE)) {
			t.push_back(std::make_pair<int, int>(x - 1, y + 0));
		}
		if (isIncluded(x + 1, y) && !isOf(x + 1, y, MSTileProperty::OBSTACLE)) {
			t.push_back(std::make_pair<int, int>(x + 1, y + 0));
		}
		
		return t;
	}

	// Checks whether the tile with the given coordinates is inside of a grid.
	bool isIncluded(int x, int y) {
		return x >= 0 && y >= 0 && x < height && y < width;
	}

	// Returns pair of coordinates of the starting tile; if multiple exist, returns coordinates for the first one.
	// If none exist, throws "No starting tile." excpetion.
	std::pair<int, int> getStart() {
		for (int i = 0; i < height; i ++) {
			for (int j = 0; j < width; j ++) {
				if (isOf(i, j, MSTileProperty::START)) {
					return std::make_pair(i, j);
				}
			}
		}

		throw MSException::NO_STARTING_TILE;
	}

	// Spawns food randomly on the board.
	void spawnFood() {
		while (true) {
			int xc = std::rand() % height;
			int yc = std::rand() % width;

			if (!isOf(xc, yc, MSTileProperty::OBSTACLE) && !isOf(xc, yc, MSTileProperty::START) && !isOf(xc, yc, MSTileProperty::MINE) && !isOf(xc, yc, MSTileProperty::SNAKE)) {
				addProperty(xc, yc, MSTileProperty::FOOD);
				break;
			}
		}
	}

	// Returns stringstream representing grid and containing each tile's properties.
	std::stringstream toStringStream() {
		std::stringstream s;

		for (auto r : grid) {
			for (auto t : r) {
				s << "[" << t.toString() << "]";
			}

			s << "\n";
		}

		return s;
	}

	// Returns string representing grid and containing each tile's properties.
	std::string toString() {
		return toStringStream().str();
	}
};

#endif
